uqcsbot/scripts/wolfram.py

```diff
@@ -0,0 +1,192 @@
+    # Recall the format of the fallback "identifier hostname s_output conversationID"
+    _, reply_host, s_output, conversation_id = parent_fallback.split(' ')
+    new_question = evt['text']  # This is the value of the message that triggered this whole response
+    s_output = '' if s_output == 'None' else s_output
+
+    # Slack annoyingly formats the reply_host link so we have to extract what we want:
+    reply_host = reply_host[1:-1].split('|')[0]
+
+    # Now we can ask Wolfram for the next answer:
+    api_url = f'{reply_host}/api/v1/conversation.jsp?'
+    params = {'appid': APP_ID, 'i': new_question, 'conversationid': conversation_id, 's': s_output}
+    http_response = await bot.run_async(requests.get, api_url, params=params)
+
+    if http_response.status_code != requests.codes.ok:
+        bot.post_message(channel, "There was a problem getting the response", thread_ts=thread_ts)
+        return
+
+    # Convert to json and check for an error
+    wolfram_answer = json.loads(http_response.content)
+    if 'error' in wolfram_answer:
+        bot.post_message(channel, wolfram_answer['error'], thread_ts=thread_ts)
+        return
+
+    # Otherwise grab the new stuff and post the reply.
+    reply = wolfram_answer['result']
+    conversation_id = wolfram_answer['conversationID']
+    reply_host = wolfram_answer['host']
+    s_output = wolfram_answer.get('s', None)
+
+    bot.post_message(channel, reply, thread_ts=thread_ts)
+
+    # Update the old fallback to reflect the new state of the conversation
+    parent_attachments['fallback'] = f'WolframCanReply {reply_host} {s_output} {conversation_id}'
+
+    bot.api.chat.update(channel=channel, attachments=[parent_attachments], ts=thread_ts)
+
+
+async def short_answer(search_query: str):
+    """
+    This uses wolfram's short answers api to just return a simple short plaintext response.
+
+    This is used if the conversation api fails to get a result (for instance !wolfram pineapple is not a great
+    conversation starter but may be interesting.
+    """
+    api_url = r"http://api.wolframalpha.com/v1/result?"
+    http_response = await bot.run_async(requests.get, api_url, params={'input': search_query, 'appid': APP_ID})
+
+    # Check if the response is ok. A status code of 501 signifies that no result could be found.
+    if http_response.status_code == 501:
+        return "No short answer available. Try !wolframfull"
+    elif http_response.status_code != requests.codes.ok:
+        return "There was a problem getting the response"
+
+    return http_response.content
+
+
+def get_subpods(pods: list) -> Iterable[Tuple[str, dict]]:
+    """
+    Yields subpods in the order they should be displayed
+
+    Yield: (pod_or_subpod_title, subpod)
+    """
+    for pod in pods:
+        for subpod in pod["subpods"]:
+            # Use the pods title if the subpod doesn't have its own title (general case)
+            title = subpod['title'] if subpod['title'] else pod['title']
```

Comment:
why not `title = subpod.get('title', pod['title'])`? Will pod always have a title?

Reply:
Are you meant to be checking if `title` exists in subpod or if its non-empty? because I think your code would KeyError if you're checking if it exists

Reply:
If you're checking its non-empty, then yeah my suggestion won't work hehe-- but I think in that case you should make your code more verbose i.e. `if len(subpod['title']) != 0`